File: win/directx/sdk-9.0/Samples/C++/Direct3D/Text3D/text3d.cpp
//-----------------------------------------------------------------------------
// File: Text3D.cpp
//
// Desc: Example code showing how to do text in a Direct3D scene.
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
#define STRICT
#include <Windows.h>
#include <commctrl.h>
#include <commdlg.h>
#include <math.h>
#include <tchar.h>
#include <stdio.h>
#include <D3DX9.h>
#include "DXUtil.h"
#include "D3DEnumeration.h"
#include "D3DSettings.h"
#include "D3DApp.h"
#include "D3DFont.h"
#include "D3DUtil.h"
#include "resource.h"

//-----------------------------------------------------------------------------
// Name: class CMyD3DApplication
// Desc: Application class. The base class (CD3DApplication) provides the 
//       generic functionality needed in all Direct3D samples. CMyD3DApplication 
//       adds functionality specific to this sample program.
//-----------------------------------------------------------------------------
class CMyD3DApplication : public CD3DApplication
{
    CD3DFont*     m_pFont;
    CD3DFont*     m_pStatsFont;
    LPD3DXMESH    m_pMesh3DText;
    LPD3DXFONT    m_pd3dxFont;

    TCHAR         m_strFont[LF_FACESIZE];
    DWORD         m_dwFontSize;

    D3DXMATRIXA16    m_matObj1;
    D3DXMATRIXA16    m_matObj2;

protected:
    HRESULT OneTimeSceneInit();
    HRESULT InitDeviceObjects();
    HRESULT RestoreDeviceObjects();
    HRESULT InvalidateDeviceObjects();
    HRESULT DeleteDeviceObjects();
    HRESULT Render();
    HRESULT FrameMove();
    HRESULT FinalCleanup();
    HRESULT CreateD3DXTextMesh( LPD3DXMESH* ppMesh, TCHAR* pstrFont, DWORD dwSize );
    HRESULT CreateD3DXFont( LPD3DXFONT* ppd3dxFont, TCHAR* pstrFont, DWORD dwSize );

public:
    LRESULT MsgProc( HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam );
    CMyD3DApplication();
};




//-----------------------------------------------------------------------------
// Name: WinMain()
// Desc: Entry point to the program. Initializes everything, and goes into a
//       message-processing loop. Idle time is used to render the scene.
//-----------------------------------------------------------------------------
INT WINAPI WinMain( HINSTANCE hInst, HINSTANCE, LPSTR, INT )
{
    CMyD3DApplication d3dApp;

    InitCommonControls();
    if( FAILED( d3dApp.Create( hInst ) ) )
        return 0;

    return d3dApp.Run();
}




//-----------------------------------------------------------------------------
// Name: CMyD3DApplication()
// Desc: Application constructor. Sets attributes for the app.
//-----------------------------------------------------------------------------
CMyD3DApplication::CMyD3DApplication()
{
    m_strWindowTitle    = _T("Text3D: Text in a 3D scene");
    m_d3dEnumeration.AppUsesDepthBuffer   = TRUE;
    m_pMesh3DText       = NULL;
    m_pd3dxFont         = NULL;

    // Create fonts
    lstrcpy( m_strFont, _T("Arial") );
    m_dwFontSize  = 18;
    m_pFont       = NULL;
    m_pStatsFont  = NULL;
}




//-----------------------------------------------------------------------------
// Name: OneTimeSceneInit()
// Desc: Called during initial app startup, this function performs all the
//       permanent initialization.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::OneTimeSceneInit()
{
    m_pFont = new CD3DFont( m_strFont, m_dwFontSize );
    if( m_pFont == NULL )
        return E_FAIL;

    m_pStatsFont = new CD3DFont( _T("Arial"), 12, D3DFONT_BOLD );
    if( m_pStatsFont == NULL )
        return E_FAIL;

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: FrameMove()
// Desc: Called once per frame, the call is the entry point for animating
//       the scene.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::FrameMove()
{
    // Setup five rotation matrices (for rotating text strings)
    D3DXVECTOR3 vAxis1(1.0f,2.0f,0.0f);
    D3DXVECTOR3 vAxis2(2.0f,1.0f,0.0f);
    D3DXMatrixRotationAxis( &m_matObj1, &vAxis1, m_fTime/2.0f  );
    D3DXMatrixRotationAxis( &m_matObj2, &vAxis2, m_fTime/2.0f  );

    // Add some translational values to the matrices
    m_matObj1._41 = 1.0f;   m_matObj1._42 = 6.0f;   m_matObj1._43 = 20.0f; 
    m_matObj2._41 = -4.0f;  m_matObj2._42 = -1.0f;  m_matObj2._43 = 0.0f; 

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: Render()
// Desc: Called once per frame, the call is the entry point for 3d
//       rendering. This function sets up render states, clears the
//       viewport, and renders the scene.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::Render()
{
    D3DMATERIAL9 mtrl;

    // Clear the viewport
    m_pd3dDevice->Clear( 0L, NULL, D3DCLEAR_TARGET | D3DCLEAR_ZBUFFER, 0x00000000, 1.0f, 0L );

    // Begin the scene 
    if( SUCCEEDED( m_pd3dDevice->BeginScene() ) )
    {
        // Draw CD3DFont in 2D (red)
        m_pFont->DrawText( 60, 100, D3DCOLOR_ARGB(255,255,0,0), 
                           _T("CD3DFont::DrawText"), D3DFONT_FILTERED );

        // Draw CD3DFont scaled in 2D (cyan)
        m_pFont->DrawTextScaled( -1.0f, 0.8f, 0.5f, // position
                                 0.1f, 0.1f,         // scale
                                 D3DCOLOR_ARGB(255,0,255,255), 
                                 _T("CD3DFont::DrawTextScaled"), 
                                 D3DFONT_CENTERED_X|D3DFONT_FILTERED );

        // Draw CD3DFont in 3D (green)
        D3DUtil_InitMaterial( mtrl, 0.0f, 1.0f, 0.0f );
        m_pd3dDevice->SetMaterial( &mtrl );
        m_pd3dDevice->SetTransform( D3DTS_WORLD, &m_matObj1 );
        m_pFont->Render3DText( _T("CD3DFont::Render3DText"), 
                              D3DFONT_CENTERED_X|D3DFONT_CENTERED_Y|D3DFONT_TWOSIDED|D3DFONT_FILTERED );

        // Draw D3DXFont mesh in 3D (blue)
        if( m_pMesh3DText != NULL )
        {
            D3DUtil_InitMaterial( mtrl, 0.0f, 0.0f, 1.0f );
            m_pd3dDevice->SetMaterial( &mtrl );
            m_pd3dDevice->SetTransform( D3DTS_WORLD, &m_matObj2 );
            m_pMesh3DText->DrawSubset(0);
        }

        // Draw D3DXFont in 2D (purple)
        RECT rc;
        SetRect( &rc, 60, 200, 0, 0 );
        m_pd3dxFont->Begin();
        m_pd3dxFont->DrawText( _T("ID3DXFont::DrawText"), -1, &rc, 
            DT_SINGLELINE | DT_CALCRECT, 0 );
        m_pd3dxFont->DrawText( _T("ID3DXFont::DrawText"), -1, &rc, 
            DT_SINGLELINE, D3DCOLOR_ARGB(255, 255, 0, 255) );
        m_pd3dxFont->End();

        // Show frame rate
        m_pStatsFont->DrawText( 2,  0, D3DCOLOR_ARGB(255,255,255,0), m_strFrameStats );
        m_pStatsFont->DrawText( 2, 20, D3DCOLOR_ARGB(255,255,255,0), m_strDeviceStats );

        // End the scene.
        m_pd3dDevice->EndScene();
    }

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: InitDeviceObjects()
// Desc: Initialize scene objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::InitDeviceObjects()
{
    HRESULT hr;

    if( FAILED( hr = m_pFont->InitDeviceObjects( m_pd3dDevice ) ) )
        return hr;

    if( FAILED( hr = m_pStatsFont->InitDeviceObjects( m_pd3dDevice ) ) )
        return hr;

    SAFE_RELEASE( m_pd3dxFont );
    if( FAILED( hr = CreateD3DXFont( &m_pd3dxFont, m_strFont, m_dwFontSize ) ) )
        return hr;

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: CreateD3DXFont()
// Desc: 
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::CreateD3DXFont( LPD3DXFONT* ppd3dxFont, 
                                           TCHAR* pstrFont, DWORD dwSize )
{
    HRESULT hr;
    LPD3DXFONT pd3dxFontNew = NULL;
    HDC hDC;
    INT nHeight;
    HFONT hFont;

    hDC = GetDC( NULL );
    nHeight = -MulDiv( dwSize, GetDeviceCaps(hDC, LOGPIXELSY), 72 );
    ReleaseDC( NULL, hDC );
    hFont = CreateFont( nHeight, 0, 0, 0, FW_DONTCARE, false, false, false, 
        DEFAULT_CHARSET, OUT_DEFAULT_PRECIS, CLIP_DEFAULT_PRECIS, DEFAULT_QUALITY,
        DEFAULT_PITCH | FF_DONTCARE, pstrFont );
    if( hFont == NULL )
        return E_FAIL;
    hr = D3DXCreateFont( m_pd3dDevice, hFont, &pd3dxFontNew );
    DeleteObject( hFont );

    if( SUCCEEDED( hr ) )
        *ppd3dxFont = pd3dxFontNew;

    return hr;
}




//-----------------------------------------------------------------------------
// Name: RestoreDeviceObjects()
// Desc: Initialize scene objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::RestoreDeviceObjects()
{
    // Restore the fonts
    m_pFont->RestoreDeviceObjects();
    m_pStatsFont->RestoreDeviceObjects();
    m_pd3dxFont->OnResetDevice();

    // Restore the textures
    m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLORARG1, D3DTA_TEXTURE );
    m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLORARG2, D3DTA_DIFFUSE );
    m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLOROP,   D3DTOP_MODULATE );
    m_pd3dDevice->SetSamplerState( 0, D3DSAMP_MINFILTER, D3DTEXF_LINEAR );
    m_pd3dDevice->SetSamplerState( 0, D3DSAMP_MAGFILTER, D3DTEXF_LINEAR );

    m_pd3dDevice->SetRenderState( D3DRS_ZENABLE, TRUE );
    m_pd3dDevice->SetRenderState( D3DRS_DITHERENABLE, TRUE );
    m_pd3dDevice->SetRenderState( D3DRS_SPECULARENABLE, TRUE );
    m_pd3dDevice->SetRenderState( D3DRS_LIGHTING, TRUE );
    m_pd3dDevice->SetRenderState( D3DRS_AMBIENT, 0x80808080);
    D3DLIGHT9 light;
    D3DUtil_InitLight(light, D3DLIGHT_DIRECTIONAL, 10.0f, -10.0f, 10.0f);
    m_pd3dDevice->SetLight(0, &light );
    m_pd3dDevice->LightEnable(0, TRUE);

    // Set the transform matrices
    D3DXVECTOR3 vEyePt    = D3DXVECTOR3( 0.0f,-5.0f,-10.0f );
    D3DXVECTOR3 vLookatPt = D3DXVECTOR3( 0.0f, 0.0f,  0.0f );
    D3DXVECTOR3 vUpVec    = D3DXVECTOR3( 0.0f, 1.0f,  0.0f );
    D3DXMATRIXA16  matWorld, matView, matProj;

    D3DXMatrixIdentity( &matWorld );
    D3DXMatrixLookAtLH( &matView, &vEyePt, &vLookatPt, &vUpVec );
    FLOAT fAspect = m_d3dsdBackBuffer.Width / (FLOAT)m_d3dsdBackBuffer.Height;
    D3DXMatrixPerspectiveFovLH( &matProj, D3DX_PI/4, fAspect, 1.0f, 100.0f );

    m_pd3dDevice->SetTransform( D3DTS_WORLD,      &matWorld );
    m_pd3dDevice->SetTransform( D3DTS_VIEW,       &matView );
    m_pd3dDevice->SetTransform( D3DTS_PROJECTION, &matProj );

    SAFE_RELEASE( m_pMesh3DText );
    if( FAILED( CreateD3DXTextMesh( &m_pMesh3DText, m_strFont, m_dwFontSize ) ) )
        return E_FAIL;

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: CreateD3DXTextMesh()
// Desc: 
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::CreateD3DXTextMesh( LPD3DXMESH* ppMesh, 
                                               TCHAR* pstrFont, DWORD dwSize )
{
    HRESULT hr;
    LPD3DXMESH pMeshNew = NULL;
    HDC hdc = CreateCompatibleDC( NULL );
    HFONT hFont;
    HFONT hFontOld;
    LOGFONT lf;

    ZeroMemory(&lf, sizeof(lf));
    lf.lfHeight = dwSize;
    lf.lfStrikeOut = 1;
    lstrcpy(lf.lfFaceName, pstrFont);
    hFont = CreateFontIndirect(&lf);
    hFontOld = (HFONT)SelectObject(hdc, hFont); 
    hr = D3DXCreateText(m_pd3dDevice, hdc, _T("D3DXCreateText"), 
        0.001f, 0.4f, &pMeshNew, NULL, NULL);
    SelectObject(hdc, hFontOld);
    DeleteObject( hFont );
    DeleteDC( hdc );
    if( SUCCEEDED( hr ) )
        *ppMesh = pMeshNew;
    return hr;
}




//-----------------------------------------------------------------------------
// Name: InvalidateDeviceObjects()
// Desc: Called when the app is exiting, or the device is being changed,
//       this function deletes any device dependent objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::InvalidateDeviceObjects()
{
    m_pFont->InvalidateDeviceObjects();
    m_pStatsFont->InvalidateDeviceObjects();
    m_pd3dxFont->OnLostDevice();

    SAFE_RELEASE( m_pMesh3DText );
    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: DeleteDeviceObjects()
// Desc: Called when the app is exiting, or the device is being changed,
//       this function deletes any device dependent objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::DeleteDeviceObjects()
{
    m_pFont->DeleteDeviceObjects();
    m_pStatsFont->DeleteDeviceObjects();
    SAFE_RELEASE( m_pd3dxFont );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: FinalCleanup()
// Desc: Called before the app exits, this function gives the app the chance
//       to cleanup after itself.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::FinalCleanup()
{
    SAFE_DELETE( m_pFont );
    SAFE_DELETE( m_pStatsFont );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: MsgProc()
// Desc: Message proc function to handle key and menu input
//-----------------------------------------------------------------------------
LRESULT CMyD3DApplication::MsgProc( HWND hWnd, UINT uMsg, WPARAM wParam,
                                    LPARAM lParam )
{
    if( WM_COMMAND == uMsg )
    {
        switch( LOWORD(wParam) )
        {
            case IDM_FONT:
            {
                HDC hdc;
                LONG lHeight;
                hdc = GetDC( hWnd );
                lHeight = -MulDiv( m_dwFontSize, GetDeviceCaps(hdc, LOGPIXELSY), 72 );
                ReleaseDC( hWnd, hdc );
                hdc = NULL;

                LOGFONT lf;
                lstrcpy( lf.lfFaceName, m_strFont );
                lf.lfHeight = lHeight;

                CHOOSEFONT cf;
                ZeroMemory( &cf, sizeof(cf) );
                cf.lStructSize = sizeof(cf);
                cf.hwndOwner   = m_hWnd;
                cf.lpLogFont   = &lf;
                cf.Flags       = CF_SCREENFONTS | CF_INITTOLOGFONTSTRUCT | CF_TTONLY;
                
                if( ChooseFont( &cf ) )
                {
                    CD3DFont* pFontNew = NULL;
                    LPD3DXMESH pMesh3DTextNew = NULL;
                    LPD3DXFONT pd3dxFontNew = NULL;
                    TCHAR* pstrFontNameNew = lf.lfFaceName;
                    DWORD dwFontSizeNew = cf.iPointSize / 10;
                    bool bSuccess = false;

                    pFontNew = new CD3DFont( pstrFontNameNew, dwFontSizeNew );
                    if( NULL != pFontNew )
                    {
                        if( SUCCEEDED( pFontNew->InitDeviceObjects( m_pd3dDevice ) ) )
                        {
                            if( SUCCEEDED( pFontNew->RestoreDeviceObjects() ) )
                            {
                                if( SUCCEEDED( CreateD3DXTextMesh( &pMesh3DTextNew, pstrFontNameNew, dwFontSizeNew ) ) )
                                {
                                    if( SUCCEEDED( CreateD3DXFont( &pd3dxFontNew, pstrFontNameNew, dwFontSizeNew ) ) )
                                    {
                                        bSuccess = true;

                                        SAFE_DELETE( m_pFont );
                                        m_pFont = pFontNew;
                                        pFontNew = NULL;

                                        SAFE_RELEASE( m_pMesh3DText );
                                        m_pMesh3DText = pMesh3DTextNew;
                                        pMesh3DTextNew = NULL;

                                        SAFE_RELEASE( m_pd3dxFont );
                                        m_pd3dxFont = pd3dxFontNew;
                                        pd3dxFontNew = NULL;
                                        
                                        lstrcpy( m_strFont, pstrFontNameNew );
                                        m_dwFontSize = dwFontSizeNew;
                                    }
                                }
                            }
                        }
                    }
                    SAFE_RELEASE( pMesh3DTextNew );
                    SAFE_RELEASE( pd3dxFontNew );
                    SAFE_DELETE( pFontNew );

                    if( !bSuccess )
                    {
                        MessageBox( m_hWnd, TEXT("Could not create that font."), 
                            m_strWindowTitle, MB_OK );
                    }
                }
                break;
            }
        }
    }

    return CD3DApplication::MsgProc( hWnd, uMsg, wParam, lParam );
}

File: win/directx/sdk-9.0/Samples/C#/Direct3D/Text3D/Text3D.cs
//-----------------------------------------------------------------------------
// File: Text3D.cs
//
// Desc: Example code showing how to do text in a Direct3D scene.
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using System.Drawing;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;




namespace Text3D
{
	/// <summary>
	/// Application class. The base class (GraphicsSample) provides the generic 
	/// functionality needed in all Direct3D samples. MyGraphicsSample adds 
	/// functionality specific to this sample program.
	/// </summary>
	public class MyGraphicsSample : GraphicsSample
	{
		private GraphicsFont drawingFont = null; // Font for drawing text
		private GraphicsFont statsFont = null; // Font for drawing text (Stats)
		private Mesh mesh3DText = null;  // Mesh to draw 3d text
		private D3DXFont otherFont = null; // The D3DX Font object

		private System.Drawing.Font ourFont = new System.Drawing.Font("Arial", 18, System.Drawing.FontStyle.Bold); // Font we use 

		private Matrix objectOne = new Matrix();
		private Matrix objectTwo = new Matrix();

		private System.Windows.Forms.MenuItem mnuOptions;
		private System.Windows.Forms.MenuItem mnuChangeFont;


        
        
        /// <summary>
		/// Application constructor. Sets attributes for the app.
		/// </summary>
		public MyGraphicsSample()
		{
			// Set the window text
			this.Text = "Text3D: Text in a 3D scene";
            try
            {
                // Load the icon from our resources
                System.Resources.ResourceManager resources = new System.Resources.ResourceManager(this.GetType());
                this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            }
            catch
            {
                // It's no big deal if we can't load our icons, but try to load the embedded one
                try { this.Icon = new System.Drawing.Icon(this.GetType(), "directx.ico"); } 
                catch {}
            }

			// Create our font objects
			statsFont = new GraphicsFont("Arial", System.Drawing.FontStyle.Bold);
			drawingFont = new GraphicsFont(ourFont);
			otherFont = new D3DXFont(ourFont);
			enumerationSettings.AppUsesDepthBuffer = true;

			// Add our new menu options
			mnuOptions = new System.Windows.Forms.MenuItem("&Options");
			mnuChangeFont = new System.Windows.Forms.MenuItem("&Change Font...");
			// Add to the main menu screen
			mnuMain.MenuItems.Add(this.mnuOptions);
			mnuOptions.MenuItems.Add(mnuChangeFont);
			mnuChangeFont.Shortcut = System.Windows.Forms.Shortcut.CtrlO;
			mnuChangeFont.ShowShortcut = true;
			mnuChangeFont.Click += new System.EventHandler(this.ChangeFontClick);
		}




        /// <summary>
        /// Handles font change
        /// </summary>
        private void ChangeFontClick(object sender, EventArgs e)
		{
			System.Windows.Forms.FontDialog dlg = new System.Windows.Forms.FontDialog();

			// Show the dialog
			dlg.FontMustExist = true;
			dlg.Font = ourFont;
			
			if (dlg.ShowDialog(this) == System.Windows.Forms.DialogResult.OK) // We selected something
			{
				ourFont = dlg.Font;
				if (drawingFont != null)
					drawingFont.Dispose(null, null);
				// Set the new font
				drawingFont = new GraphicsFont(ourFont);
				drawingFont.InitializeDeviceObjects(device);

				if (otherFont != null)
					otherFont.Dispose();
				otherFont = new D3DXFont(ourFont);
				otherFont.InitializeDeviceObjects(device);

				// Create our 3d text mesh
				mesh3DText = Mesh.TextFromFont(device, ourFont, "Mesh.TextFromFont", 0.001f, 0.4f);
			}
		}




		/// <summary>
		/// Called once per frame, the call is the entry point for animating the scene.
		/// </summary>
		protected override void FrameMove()
		{
			// Setup five rotation matrices (for rotating text strings)
			Vector3 vAxis1 = new Vector3(1.0f,2.0f,0.0f);
			Vector3 vAxis2 = new Vector3(2.0f,1.0f,0.0f);
			objectOne = Matrix.RotationAxis(vAxis1, appTime / 2.0f);
			objectTwo = Matrix.RotationAxis(vAxis2, appTime / 2.0f);

			// Add some translational values to the matrices
			objectOne.M41 = 1.0f;   objectOne.M42 = 6.0f;   objectOne.M43 = 20.0f; 
			objectTwo.M41 = -4.0f;  objectTwo.M42 = -1.0f;  objectTwo.M43 = 0.0f; 
		}




		/// <summary>
		/// Called once per frame, the call is the entry point for 3d rendering. This 
		/// function sets up render states, clears the viewport, and renders the scene.
		/// </summary>
		protected override void Render()
		{
			// Clear the viewport
			device.Clear(ClearFlags.Target | ClearFlags.ZBuffer, Color.Black, 1.0f, 0);

			device.BeginScene();

			// Output statistics
			statsFont.DrawText(2,  1, Color.Yellow, frameStats);
			statsFont.DrawText(2, 20, Color.Yellow, deviceStats);

			// Draw GraphicsFont in 2D (red)
			drawingFont.DrawText(60, 100, Color.Red, "GraphicsFont.DrawText");

			// Draw GraphicsFont scaled in 2D (cyan)
			drawingFont.DrawTextScaled(-1.0f, 0.8f, 0.5f, // position
								        0.1f, 0.1f,       // scale
								        Color.Cyan, "GraphicsFont.DrawTextScaled");

			// Draw GraphicsFont in 3D (green)
			Material mtrl = GraphicsUtility.InitMaterial(Color.Green);
			device.Material = mtrl;
			device.Transform.World = objectOne;
			drawingFont.Render3DText("GraphicsFont.Render3DText", GraphicsFont.RenderFlags.Centered | GraphicsFont.RenderFlags.TwoSided | GraphicsFont.RenderFlags.Filtered);

			// Draw D3DXFont mesh in 3D (blue)
			if (mesh3DText != null)
			{
				Material mtrl3d = GraphicsUtility.InitMaterial(Color.Blue);
				device.Material = mtrl3d;
				device.Transform.World = objectTwo;
				mesh3DText.DrawSubset(0);
			}

			// Draw D3DXFont in 2D (purple)
			otherFont.BeginText();
			otherFont.DrawText(60, 200, Color.Purple.ToArgb(), "D3DXFont.DrawText");
			otherFont.EndText();

			device.EndScene();
		}




		/// <summary>
        /// The device has been created.  Resources that are not lost on
        /// Reset() can be created here -- resources in Pool.Managed,
        /// Pool.Scratch, or Pool.SystemMemory.  Image surfaces created via
        /// CreateImageSurface are never lost and can be created here.  Vertex
        /// shaders and pixel shaders can also be created here as they are not
        /// lost on Reset().
		/// </summary>
		protected override void InitializeDeviceObjects()
		{
			// Initialize all of the fonts
			drawingFont.InitializeDeviceObjects(device);
			statsFont.InitializeDeviceObjects(device);
			otherFont.InitializeDeviceObjects(device);
		}




		/// <summary>
        /// The device exists, but may have just been Reset().  Resources in
        /// Pool.Default and any other device state that persists during
        /// rendering should be set here.  Render states, matrices, textures,
        /// etc., that don't change during rendering can be set once here to
        /// avoid redundant state setting during Render() or FrameMove().
		/// </summary>
		protected override void RestoreDeviceObjects(System.Object sender, System.EventArgs e)
		{
			// Restore the textures
			device.TextureState[0].ColorArgument1 = TextureArgument.TextureColor;
			device.TextureState[0].ColorArgument2 = TextureArgument.Diffuse;
			device.TextureState[0].ColorOperation = TextureOperation.Modulate;
			device.SamplerState[0].MinFilter = TextureFilter.Linear;
			device.SamplerState[0].MagFilter = TextureFilter.Linear;

			device.RenderState.ZBufferEnable = true;
			device.RenderState.DitherEnable = true;
			device.RenderState.SpecularEnable = true;
			device.RenderState.Lighting = true;
			device.RenderState.Ambient = System.Drawing.Color.FromArgb(unchecked((int)0x80808080));

			GraphicsUtility.InitLight(device.Lights[0], LightType.Directional, 10.0f, -10.0f, 10.0f);
			device.Lights[0].Commit();
			device.Lights[0].Enabled = true;

			// Set the transform matrices
			Vector3 vEyePt = new Vector3(0.0f,-5.0f,-10.0f);
			Vector3 vLookatPt = new Vector3(0.0f, 0.0f,  0.0f);
			Vector3 vUpVec = new Vector3(0.0f, 1.0f,  0.0f);
			Matrix matWorld, matView, matProj;

			matWorld = Matrix.Identity;
			matView = Matrix.LookAtLH(vEyePt, vLookatPt, vUpVec);
			float fAspect = device.PresentationParameters.BackBufferWidth / (float)device.PresentationParameters.BackBufferHeight;
			matProj = Matrix.PerspectiveFovLH((float)Math.PI / 4, fAspect, 1.0f, 100.0f);

			device.Transform.World = matWorld;
			device.Transform.View = matView;
			device.Transform.Projection = matProj;

			if (mesh3DText != null)
				mesh3DText.Dispose();

			// Create our 3d text mesh
			mesh3DText = Mesh.TextFromFont(device, ourFont, "Mesh.TextFromFont", 0.001f, 0.4f);
		}




        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        static void Main() 
        {
            using (MyGraphicsSample d3dApp = new MyGraphicsSample())
            {                                 
                if (d3dApp.CreateGraphicsSample())
                    d3dApp.Run();
            }
        }
    }
}
